Step3  - QnA RESTful API 기능구현

willing1026 opened the pull request:
Repository없이 AnswerAcceptanceTest를 완성하고 싶었는데,
계속해서 제가 알지 못하는 오류들이 발생하여 결국 Respository를 사용했습니다.

그리고 AnswerResponseDto를 만든이유도 json Parsing 오류와 같은 문제들.
커넥션이 끊어졌다는 오류 등 알지못하는 부분에 대한 오류들이 많아서 만들게 됐습니다.

모르는게 너무 많네요..  피드백 부탁드립니다!

willing1026 committed:
- QuestionTest 수정. ( Repository 기능점검 내용 -> QuestionRepositoryTest 클래스로 분리 ,  QuestionTest는 기본 Junit 테스트로 작성 )

willing1026 committed:
# step3 - ApiQuestionController 구현 및 테스트

#FIX
- QuestionAcceptanceTest : 잘못된 접근에 대한 Exception 변경 (UnAuthorizedException) , 주석제거
- QnaService : 예외처리 수정 (UnAuthorizedException 발생하도록 변경)
- QuestionController : Exception처리 변경으로 try-catch 구문 제거
- Question : Exception수정, ApiUri generate메서드 추가
- QuestionTest : 수정, 삭제에 대한 TestCase수정
- AcceptanceTest : QuestionRepository 사용제거

#ADD
- ApiQuestionAcceptance : Question Rest API Controller 추가
- ApiQuestionAcceptanceTest : Controller Acceptance Test 추가
- AcceptanceTest : Question자원 post, get 에 대한 공통기능 메서드로 분리. 중복제거 ( createResource, getResoruce )

willing1026 committed:
- method 순서 정리

willing1026 committed:
# Answer Entity update, delete 메서드 추가 및 테스트

willing1026 committed:
# 코드 정리

willing1026 committed:
# 패키지 정리

willing1026 committed:
#step3 - QnA CRUD 기능 구현, 테스트

- AnswerResponseDto 객체 생성, Answer 에 대한 Response는 Dto객체로 변환, 전달하도록 개발

willing1026 committed:
# 미사용코드 제거

kang-hyungu reviewed (changes requested):
안녕하세요 선호님
몇 가지 수정하실 사항들 피드백 남겼어요.
그리고 getter 때문에 무한 n+1이 발생하는 문제가 있네요

1. 답변 수정하는 API가 없네요. 수정 기능도 추가해보아요
2. 그리고 AnswerResponseDto를 안쓰고 도메인을 바로 호출하려면 Answer 클래스에서 getQuestion 메서드를 제거해주세요.
/api/questions/1/answers/1을 호출할 때 jackson에서 파싱할때 getter를 사용하는데, getQuestion이 호출되면서 answer의 question을 매핑하고 다시 question에 answer를 매핑하고 answer의 question이 매핑되는... 무한 매핑 지옥에 빠져서 생긴 문제입니다.
불필요한 getter는 추가하지 않도록 합시다.
3. 위 이슈를 해결하고 다시 repository test를 사용하지 않도록 수정해보아요!

willing1026 committed:
#피드백 반영 (N+1 문제 개선, 테스트 수정, 답변 수정 기능 추가, 코드정리)

- Test에서 Repository 호출 제거, API 호출을 통한 인수테스트
- 미사용 코드 제거. 코드개선
- 답변 수정기능 추가, 테스트
- DTO 안쓰도록 개선 ( @Deprecated 어노테이션 등록 )

kang-hyungu commented on src/main/java/nextstep/service/QnaService.java:
```diff
@@ -64,12 +83,35 @@ public void deleteQuestion(User loginUser, long questionId) throws CannotDeleteE
+    }
+
+    @Transactional(readOnly = true)
+    public AnswerResponseDto showAnswer(long questionId, long answerId) {
```
showAnswer 메서드는 단순 답변 조회라서 저라면 answerId로 조회할거에요
answer 조회하면 jpa가 자동으로 매핑된 question을 가져올거라 question id 유효성 체크가 의미가 없습니다

willing1026 replied:
오.. 감사합니다

kang-hyungu commented on src/test/java/nextstep/domain/QnaRepositoryTest.java:
```diff
@@ -0,0 +1,84 @@
+
+@RunWith(SpringRunner.class)
+@DataJpaTest
+public class QnaRepositoryTest {
```
Deprecated 어노테이션은 삭제 예정이라는 의미라 주석으로 학습 테스트 코드라고 남겨주시면 협업할 때 충분히 의미 전달이 될거에요

willing1026 replied:
네 감사합니다!!

kang-hyungu commented on src/test/java/nextstep/web/ApiAnswerAcceptanceTest.java:
```diff
@@ -0,0 +1,81 @@
+
+    @Test
+    public void delete() {
+        Answer newAnswer = answerRepository.findById(1L).orElseThrow(UnAuthorizedException::new);
```
question이 null로 나왔던 테스트 코드를 올려주시겠어요?
질문 내용만으로는 상황을 알기가 어렵네요

kang-hyungu reviewed (approved):
피드백 반영 확인했습니다
코드가 이전보다 더 클린 코드가 됐네요
3단계 pr은 머지할게요.
마지막 단계 진행하시기 바랍니다! 수고하셨습니다.

그리고 질문 주신 내용은 코멘트 남겼어요.
api 테스트 할 때 question 객체가 null로 나오는 케이스는 작성하셨던 코드를 남겨주시거나 좀더 상세한 상황을 설명해주시면 확인해볼게요. 남겨주신 내용으로는 어떤 상황인지 파악이 어렵네요.

willing1026 commented on src/test/java/nextstep/domain/AnswerTest.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Test
+    public void mapping_question() {
```
question 객체에 생성자 추가 (id, title, contents, user)
answer 객체에서 questionId get메서드 추가
이것들 활용해서 테스트 수정
